src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,469 @@
+            this.currentBackupVersionCheck = null;
+        }
+    }
+
+    private async internalCheckFromServer(): Promise<Configuration | null> {
+        let currentServerVersion = null;
+        try {
+            currentServerVersion = await this.backupManager.requestKeyBackupVersion();
+        } catch (e) {
+            this.logger.debug(`Backup: error while checking server version: ${e}`);
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+        this.logger.debug(`Got current version from server:${currentServerVersion?.version}`);
+
+        if (currentServerVersion?.algorithm != "m.megolm_backup.v1.curve25519-aes-sha2") {
+            this.logger.info(`getBackupDecryptor Unsupported algorithm ${currentServerVersion?.algorithm}`);
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        if (!currentServerVersion?.version) {
+            this.logger.info(`No current key backup`);
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        const activeVersion = await this.backupManager.getActiveBackupVersion();
+        if (activeVersion == null || currentServerVersion.version != activeVersion) {
+            // case when the server side current version is not trusted or is out of sync with the client side active version.
+            this.logger.info(
+                `The current backup version ${currentServerVersion.version} is not trusted. Active version=${activeVersion}`,
+            );
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        const authData = <Curve25519AuthData>currentServerVersion.auth_data;
+
+        const backupKeys = await this.getBackupDecryptionKey();
+        if (!backupKeys?.decryptionKey) {
+            this.logger.debug(`Not checking key backup for session(no decryption key)`);
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        if (activeVersion != backupKeys.backupVersion) {
+            this.logger.debug(
+                `Cached key version <${backupKeys.backupVersion}> doesn't match active backup version <${activeVersion}>`,
+            );
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        if (authData.public_key != backupKeys.decryptionKey.megolmV1PublicKey.publicKeyBase64) {
+            this.logger.debug(`getBackupDecryptor key mismatch error`);
+            this.hasConfigurationProblem = true;
+            return null;
+        }
+
+        const backupDecryptor = this.backupManager.createBackupDecryptor(backupKeys.decryptionKey);
+        this.hasConfigurationProblem = false;
+        this.configuration = {
+            decryptor: backupDecryptor,
+            backupVersion: activeVersion,
+        };
+        return this.configuration;
+    }
+
+    private async decryptAndImport(sessionInfo: SessionInfo, data: KeyBackupSession): Promise<void> {
```

Comment:
sure why not. I have no idea of the expectations regarding method orders in fi;es

Reply:
> I have no idea of the expectations regarding method orders in files

I'm just looking for some sort of logic, rather than what appears to be a completely random ordering